homeassistant/components/sensor/recorder.py

```diff
@@ -272,9 +291,28 @@ def compile_statistics(
 
             for fstate, state in fstates:
 
-                if "last_reset" not in state.attributes:
+                # Deprecated, will be removed in Home Assistant 2021.10
+                if (
+                    "last_reset" not in state.attributes
+                    and state_class == STATE_CLASS_MEASUREMENT
+                ):
                     continue
-                if (last_reset := state.attributes["last_reset"]) != old_last_reset:
+
+                reset = False
+                if (
+                    state_class != STATE_CLASS_TOTAL_INCREASING
+                    and (last_reset := state.attributes.get("last_reset"))
+                    != old_last_reset
+                ):
+                    reset = True
+                elif old_state is None and last_reset is None:
+                    reset = True
+                elif state_class == STATE_CLASS_TOTAL_INCREASING and (
+                    old_state is None or fstate < old_state
```

Comment:
Might be worth usings omething like `fstate < 0.9 * old_state`? I'm just thinking of cases where a slightly noisy energy sensor is being used in an integration to calculate. If the sensor is measuring 0 +/- some noise it is possible for the value to creep down a small amount without being a true reset

Reply:
I think that needs to be handled by the integration if it claims `STATE_CLASS_TOTAL_INCREASING`

Reply:
Agreed. I think it's best that the integration claiming that the sensor will only increase ensures that is true. I thought of some others where a less than could be an issue but in all cases the integration should know what it's doing better